lib/logstash/outputs/datadog_logs.rb

```diff
@@ -6,67 +6,231 @@
-  config :use_ssl,     :validate => :boolean, :required => true, :default  => true
-  config :max_backoff, :validate => :number,  :required => true, :default  => 30
-  config :max_retries, :validate => :number,  :required => true, :default  => 5
+  config :api_key, :validate => :string, :required => true
+  config :host, :validate => :string, :required => true, :default => "http-intake.logs.datadoghq.com"
+  config :port, :validate => :number, :required => true, :default => 443
+  config :use_ssl, :validate => :boolean, :required => true, :default => true
+  config :max_backoff, :validate => :number, :required => true, :default => 30
+  config :max_retries, :validate => :number, :required => true, :default => 5
+  config :use_http, :validate => :boolean, :required => false, :default => true
+  config :use_compression, :validate => :boolean, :required => false, :default => true
+  config :compression_level, :validate => :number, :required => false, :default => 6
+  config :no_ssl_validation, :validate => :boolean, :required => false, :default => false
 
+  # Register the plugin to logstash
   public
   def register
-    require "socket"
-    client = nil
-    @codec.on_event do |event, payload|
-      message = "#{@api_key} #{payload}\n"
-      retries = 0
+    @client = new_client(@logger, @api_key, @use_http, @use_ssl, @no_ssl_validation, @host, @port, @use_compression)
+  end
+
+  # Entry point of the plugin, receiving a set of Logstash events
+  public
+  def multi_receive(events)
+    return if events.empty?
+    encoded_events = @codec.multi_encode(events)
+    if @use_http
+      batches = batch_http_events(encoded_events, DD_MAX_BATCH_LENGTH, DD_MAX_BATCH_SIZE)
+      batches.each do |batched_event|
+        process_encoded_payload(format_http_event_batch(batched_event))
+      end
+    else
+      encoded_events.each do |encoded_event|
+        process_encoded_payload(format_tcp_event(encoded_event.last, @api_key, DD_MAX_BATCH_SIZE))
+      end
+    end
+  end
+
+  # Process and send each encoded payload
+  def process_encoded_payload(payload)
+    if @use_compression and @use_http
+      payload = gzip_compress(payload, @compression_level)
+    end
+    @client.send_retries(payload, @max_retries, @max_backoff)
+  end
+
+  # Format TCP event
+  def format_tcp_event(payload, api_key, max_request_size)
+    formatted_payload = "#{api_key} #{payload}"
+    if (formatted_payload.bytesize > max_request_size)
+      return truncate(formatted_payload, max_request_size)
+    end
+    formatted_payload
+  end
+
+  # Format HTTP events
+  def format_http_event_batch(batched_events)
+    "[#{batched_events.join(',')}]"
+  end
+
+  # Group HTTP events in batches
+  def batch_http_events(encoded_events, max_batch_length, max_request_size)
+    batches = []
+    current_batch = []
+    current_batch_size = 0
+    encoded_events.each_with_index do |event, i|
+      encoded_event = event.last
```

Comment:
I'm not sure to understand this line, why `last` returns the `encoded_event`?

Reply:
Good question. Logstash returns both the original event and the encoded event after being encoded with @codec. So last is here to get the encoded event.